plugin/pkg/managedseed/validator/admission.go

```diff
@@ -0,0 +1,373 @@
+		return New()
+	})
+}
+
+// ValidateManagedSeed contains listers and and admission handler.
+type ValidateManagedSeed struct {
+	*admission.Handler
+	coreClient coreclientset.Interface
+	kubeClient kubernetes.Interface
+	readyFunc  admission.ReadyFunc
+}
+
+var (
+	_ = admissioninitializer.WantsInternalCoreClientset(&ValidateManagedSeed{})
+	_ = admissioninitializer.WantsKubeClientset(&ValidateManagedSeed{})
+
+	readyFuncs = []admission.ReadyFunc{}
+)
+
+// New creates a new ValidateManagedSeed admission plugin.
+func New() (*ValidateManagedSeed, error) {
+	return &ValidateManagedSeed{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+	}, nil
+}
+
+// AssignReadyFunc assigns the ready function to the admission handler.
+func (v *ValidateManagedSeed) AssignReadyFunc(f admission.ReadyFunc) {
+	v.readyFunc = f
+	v.SetReadyFunc(f)
+}
+
+// SetInternalCoreClientset sets the garden core clientset.
+func (v *ValidateManagedSeed) SetInternalCoreClientset(c coreclientset.Interface) {
+	v.coreClient = c
+}
+
+// SetKubeClientset gets the clientset from the Kubernetes client.
+func (v *ValidateManagedSeed) SetKubeClientset(c kubernetes.Interface) {
+	v.kubeClient = c
+}
+
+// ValidateInitialization checks whether the plugin was correctly initialized.
+func (v *ValidateManagedSeed) ValidateInitialization() error {
+	if v.coreClient == nil {
+		return errors.New("missing garden core client")
+	}
+	if v.kubeClient == nil {
+		return errors.New("missing kubernetes client")
+	}
+	return nil
+}
+
+var _ admission.MutationInterface = &ValidateManagedSeed{}
+
+// Admit validates and if appropriate mutates the given managed seed against the shoot that it references.
+func (v *ValidateManagedSeed) Admit(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	// Wait until the caches have been synced
+	if v.readyFunc == nil {
+		v.AssignReadyFunc(func() bool {
+			for _, readyFunc := range readyFuncs {
+				if !readyFunc() {
+					return false
+				}
+			}
+			return true
+		})
+	}
+	if !v.WaitForReady() {
+		return admission.NewForbidden(a, errors.New("not yet ready to handle request"))
+	}
+
+	// Ignore all kinds other than Shoot
+	if a.GetKind().GroupKind() != seedmanagement.Kind("ManagedSeed") {
+		return nil
+	}
+
+	// Ignore updates to status or other subresources
+	if a.GetSubresource() != "" {
+		return nil
+	}
+
+	// Convert object to ManagedSeed
+	managedSeed, ok := a.GetObject().(*seedmanagement.ManagedSeed)
+	if !ok {
+		return apierrors.NewBadRequest("could not convert object to ManagedSeed")
+	}
+
+	// Get shoot
+	shoot, err := v.coreClient.Core().Shoots(managedSeed.Namespace).Get(ctx, managedSeed.Spec.Shoot.Name, clientkubernetes.DefaultGetOptions())
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return apierrors.NewBadRequest(fmt.Sprintf("shoot %s/%s not found", managedSeed.Namespace, managedSeed.Spec.Shoot.Name))
```

Comment:
Can you use `NewInvalid` instead of `NewBadRequest` in all cases like this?
I think, the request itself is not bad or invalid but rather the specified object, thus it should be semantically cleaner to return an invalid response, I guess.
```suggestion
			return apierrors.NewInvalid(fmt.Sprintf("shoot %s/%s not found", managedSeed.Namespace, managedSeed.Spec.Shoot.Name))
```

Reply:
Yes, you are right this is better. Interestingly, `NewInvalid` is not used at all in all admission plugins right now (0 occurrences), instead `NewBadRequest` and `NewForbidden` (even worse) are used in similar situations. Not sure if we want to do something about it.

Reply:
Ups 😄 
Then you can use the proper errors here (i.e. `NewInvalid`) and open an issue or even PR to set things straight in the other plugins. Otherwise it will get lost again, though not a dramatic inconsistency and rather a cosmetic thing.

Reply:
OK, I will use `NewInvalid` and open an issue (or PR) for the rest.